### PR TITLE
New tutorial on Recurrent Neural Networks (RNN)

### PR DESCRIPTION
**Pull request by kxk302**
This is a new tutorial for Statistics and Machine Learning topic. I have uploaded the data to Zenodo, and have included the workflow as well. I followed the instructions/tutorials on GTN to write this tutorial. Please review and let me know if any corrections are needed. Thank you! 

P.S. As suggested, I used yEd to create the figures in this tutorial. Was wondering if its ok to add the .graphml files in the images folder (or a new folder), in case the figures need to be updated later?

### CODE REVIEW

**Comment by hexylena** [comment]
@anuprulez as one of our newest maintainers would you have a chance to look at this? Is the content ok, is everything explained well enough for beginners, etc?

**Comment by anuprulez** [comment]
@hexylena Yes, I will have a look at this tutorial a bit later today. Thanks!

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:12]: Maybe add one more objective: `Learn how to create a neural network using Galaxy's deep learning tools`

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:35]: `data` should be sufficient instead of `training data` because the latter leads towards the concept of `training/test data`

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:39]: An explanation should be there why RNN is used for sentiment analysis problem (because RNN is generally used for temporal data such as text, speech and so on ...). Please also add some successful examples (links to papers) where RNN is used for analysis of temporal data to provide more context to readers. Thanks.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:55]: Please explain what are training example fields. Do you mean "features"?

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:68]: It would be nice to have links to these activation functions such as sigmoid, tanh and so on for further reading.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:86]: Please explain what a training set is. Why it is used this way and include the significance of test set as well.

**Review by anuprulez — CHANGES_REQUESTED** [review]

**Comment by anuprulez** [comment]
@kxk302 thanks for the tutorial. I have added some comments and will add some more.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:35]: Sure,

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:39]: Will do.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:55]: I'll re-word this so its more clear.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:68]: Added a reference to a paper that reviews activation functions.

**Comment by kxk302** [comment]
> @kxk302 thanks for the tutorial. I have added some comments and will add some more.

Thanks @anuprulez. I revised the tutorial based on your comment and pushed it to my branch.

**Review by anuprulez — CHANGES_REQUESTED** [review]

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (anuprulez) [inline topics/statistics/tutorials/RNN/tutorial.md:350]: I think a step to include a dataset of `predicted labels` is missing. Plotting a confusion matrix would require true and predicted both datasets.

**Review by anuprulez — CHANGES_REQUESTED** [review]

**Comment by anuprulez** [comment]
@kxk302 Thanks for updating the tutorial. I tried running the tutorial locally but I ran into some problems. Can you address it? This is my history: https://usegalaxy.eu/u/kumara/h/rnn-tutorial

Thanks for working on it :)

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:140]: Fixed.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:219]: I have reduced the size by throwing away the top 50 most frequent words, only considering the next 10,000 most frequent words, and restricting reviews to 500 words (any longer review will be trimmed). I tried to reduce the size further by using a subset of the rows in training/test data, but it affected the performance of the RNN badly. Right now the training is pretty fast, so I think we are fine.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:359]: Will add another paragraph on how to interpret the confusion matrix.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:146]: Sure

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:253]: Good idea.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:265]: Added a sentence to explain what it is.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:288]: Deleted. Thx

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:230]: Will update accordingly.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:350]: Added the step.

**Comment by kxk302** [comment]
> https://usegalaxy.eu/u/kumara/h/rnn-tutorial

Added name/label to all inputs. Looking at workflow issue next.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:12]: Added this.

**Comment by kxk302** [comment]
@anuprulez I imported the latest workflow and it completed successfully. The only thing I had to do was convert y_test and y_train to tabular (Updated the instructions so its clear).

**Comment by kxk302** [comment]
Thank you for all the helpful feedback!

**Comment by shiltemann** [comment]
@kxk302 this is very cool, thanks! With respect to your question, yes, please add the .graphml source file in case we want to update the image later :+1:

**Comment by kxk302** [comment]
> @kxk302 this is very cool, thanks! With respect to your question, yes, please add the .graphml source file in case we want to update the image later 👍

Will add .graphml files. Thanks @shiltemann!

**Comment by hexylena** [comment]
Build failure is due to #2333

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:248]: If you could use the new tool linking syntax it would be really nice for users!

We have a [guide on how to do it](https://training.galaxyproject.org/training-material/topics/contributing/tutorials/create-new-tutorial-content/tutorial.html#tool-links), but in short just switch to `{% tool [Create a deep learning model architecture](toolshed.g2.bx.psu.edu/repos/..../your-tool/1.0.0+galaxy0) %}`

It can also be done for the upload steps above.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:322]: All of these post-tool comment boxes probably don't need to be a comment box.

```suggestion
The training step generates 3 datasets. 1) accuracy of the trained model, 2) the trained model, downloadable as a zip file, and 3) the trained 
  model weights, downloadable as an hdf5 file. These files are needed for prediction in the next step. 
```

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md]: @hexylena I added the new tool linking syntax to line 248. But after running 'make serve' and viewing the page in a browser, I just see a link that does not work. Could you please take a look at the code on line 248 and see if you see any issues? BTW, I tried grep'ing the repo to find another example, but could not find one.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md]: Moved them all out of comment sections.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md]: FYI, I updated all instances to use new toll linking syntax.

Comment (hexylena) [reply topics/statistics/tutorials/RNN/tutorial.md]: Here's another example of it being used: https://github.com/galaxyproject/training-material/blob/master/topics/visualisation/tutorials/circos/tutorial.md, there are a couple, we're slowly updating things :)

The link is *expected* to not work when you're just viewing the training (hovering over it should give the tool ID). It only is active when viewed from within Galaxy.

Your changes look exactly correct, I can check locally later.

**Comment by kxk302** [comment]
@anuprulez and @hexylena,
Was wondering if there is any other change that you suggest to the tutorial? (if not, can this be merged?)
Thanks!

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:65]: Could you please supply some more detailed alt-text here? Here's an example of what is appropriate in each field.

Field | Example
--- | --- 
alt text |  Image of cell membrance with an embedded protein with central pore. DNA is shown splitting and entering the pore, an electrical signal comes out reading A C T or G.
fig caption | Using nanopore sequencing, a single molecule of DNA or RNA can be sequenced without the need for PCR amplification or chemical labeling of the sample. (Image from: <a href="https://nanoporetech.com/sites/default/files/s3/white-papers/WGS_Assembly_white_paper.pdf?submissionGuid=40a7546b-9e51-42e7-bde9-b5ddef3c3512">Nanopore sequencing: The advantages of long reads for genome assembly</a>)

And some more comments here https://training.galaxyproject.org/training-material/topics/contributing/tutorials/create-new-tutorial-content/tutorial.html#guidelines-on-alt-vs-figcaption-text

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:175]: I'm slightly worried about the figure references (not so many others do this) if you want to e.g. add a picture at the top, but we can fix that when it becomes an issue maybe.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:218]:
```suggestion
>    - Import the files from [Zenodo](https://zenodo.org/record/4477881) or from the shared data library
```

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:235]:
```suggestion
>    {% include snippets/change_datatype.md datatype="tabular" %}
```

that datatype var is used in the snippet, and should match your desired target datatype.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [inline topics/statistics/tutorials/RNN/tutorial.md:357]: If you want to represent mathematical things you can do something like this, I think. (please test locally how it renders.)

```suggestion
to use the **F score**, which is the harmonic mean of precision and recall 

$$ F = \frac{2 * \text{precision * recall}}{\text{precision + recall}} $$
```

https://training.galaxyproject.org/training-material/topics/contributing/tutorials/create-new-tutorial-content/tutorial.html#writing-mathematical-expressions

**Review by hexylena — COMMENTED** [review]
This is looking very good from the formatting side (I know nothing about ML), I've made some comments and after these it's good to merge. (Feel free to skip the last / mathematical presentation one, it's not important. the others more.)

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (hexylena) [reply topics/statistics/tutorials/RNN/tutorial.md:65]: This is very important for accessibility, so vision impaired and blind users can access the content as well.

**Comment by anuprulez** [comment]
I will run it in my local this week and provide comments if any. Thanks @kxk302

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:65]: Added alt text to all figures in the tutorial.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:218]: Done

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:175]: Ok, so I leave it as is for now.

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:235]: Done

File: topics/statistics/tutorials/RNN/tutorial.md
Comment (kxk302) [reply topics/statistics/tutorials/RNN/tutorial.md:357]: Thanks. The formula renders fine locally.

**Comment by kxk302** [comment]
> I will run it in my local this week and provide comments if any. Thanks @kxk302

Thanks @anuprulez!

**Comment by kxk302** [comment]
@hexylena I pushed my changes for your comments.

**Review by hexylena — APPROVED** [review]
Perfect! I'm happy with this from the formatting/structure side. Over to Anup to merge when he's happy!

**Comment by anuprulez** [comment]
@kxk302 thanks for making the changes. I checked the tutorial in my local and all the steps are working fine. However, I have one minor suggestion - I do not see the final confusion matrix plot in the tutorial. Do you think it would be good to add this plot and write one or two sentences about predictions/plot?

Apart from this, it looks good to me :+1:

**Comment by kxk302** [comment]
> @kxk302 thanks for making the changes. I checked the tutorial in my local and all the steps are working fine. However, I have one minor suggestion - I do not see the final confusion matrix plot in the tutorial. Do you think it would be good to add this plot and write one or two sentences about predictions/plot?
> 
> @hexylena Apart from this, it looks good to me 👍

@anuprulez I think its a good idea to add the confusion matrix with some explanation. Will add it and update this ticket when its pushed.

**Comment by kxk302** [comment]
@anuprulez I just pushed my change that adds a Confusion Matrix image and analysis.

**Comment by anuprulez** [comment]
@kxk302 @hexylena :+1:

**Review by anuprulez — APPROVED** [review]

**Comment by hexylena** [comment]
Thanks so much @kxk302! It'll be online in ~5-10 minutes.

**Comment by kxk302** [comment]
@anuprulez @hexylena Thank you both very much for all the constructive feedback!
Anup, I have a couple more ML tutorials in the pipeline so we will meet again soon :-)

**Comment by bgruening** [comment]
That is pretty cool! Thanks @kxk302 and @anuprulez!

**Comment by anuprulez** [comment]
@kxk302 thanks :)

**Pull request merged**
